### ISSUE TITLE
Map-style vs Iterable-style Dataset

### ISSUE THREAD

**Issue by pzelasko** [lhotse-speech/lhotse#113]
I want to start a discussion to determine what's the right way to go forward with data loading efficiency and flexbility for Lhotse PyTorch datasets.

[PyTorch documentation for dataset API](https://pytorch.org/docs/stable/data.html) describes two types of datasets - map-style and iterable-style. Short summary:

### Map-style:

```python
for index in sampler:
    sample = dataset[index]
```

- loads a single item at a time
- `DataLoader` samples indices of the items to be loaded (using a `Sampler`) based on `len()` of the dataset
- `DataLoader` specifies a fixed batch size and takes care of collation (possibly with a custom `collate_fn`)

### Iterable-style:

```python
for batch in iter(dataset):  # batch/sample depending on the implementation
    pass
```

- may load multiple items at the same time
- has to care of collation, shuffling, batch size (can be dynamic), etc. itself
- does not have to specify `len()`

### Comments 

Our current examples are based on the map-style paradigm, which offloads more things to the DataLoader default settings. I'm wondering if we should explore the iterable-style dataset more going forward. It seems like it might make it easier to work with sharded data, or support non-random-access reading (e.g. block-random sampling for batches). I think it'd be good to start collecting some insights, experiences and requirements to make sure the design is solid.

Another question is should Lhotse concern itself with that, and if so, to what extent? I don't think we'll be able to provide an optimal solution for every use case, but I think it'd be good to provide an option that (perhaps with some tuning) is at least okay in most compute infrastructures. We'll also gain more insights regarding that from building and running K2 recipes.

**Comment by qindazhu** [comment]
+1.

I believe iterable-style is very useful in scenario where the data is arriving as as part of a stream or the users can not map each data point into an index, e.g. they cannot load all the data into memory, or even cannot figure out the `len()`  without loading data as the data source is a stream or a very large file, they also have no meta-file like scp file in Kaldi. 
I'm also wondering if there's any scenario that the data may come from different sources and we need to combine them together on-the-fly? BTW, handling source on a remote server is much complex, we may also need to support, e.g. read throughput (instead of read a single file each time as that will be too slow), I'm not sure how much effort we should make about this? or just for 1-2 specific remote storage format?

BTW, do we have any plan to support prefetch data? i.e. prefetch data for the next few batches while training with the current batch.

**Comment by danpovey** [comment]
But the way lhotse works is to pre-compute manifests of everything and only
the actual computation happens on the fly.
It seems to me that that is compatible with indexed loading.  Might be
easier to implement.
I think for prefetch, we might deal with that by just using a
multi-threaded loader which I think PyTorch already supports.

**Comment by pzelasko** [comment]
Dan is correct in that Lhotse "assumes" we "know" about all the data at our disposal, and we have it described in manifests. Eventually, we will extend Lhotse so it will be possible to download it on-the-fly from a remote server too (it shouldn't be too much work).

**Comment by pzelasko** [comment]
A related topic that came up in a separate discussion:

### Padding

It's one of the reasons I'm looking at iterable datasets - they would allow us to select N cuts and call pad() on just them as opposed to the whole set. Then we could also use e.g. noise for padding instead of silence/low energies. We could easily adjust the batch size dynamically based on the total duration constraints.

The alternative is to sort all cuts by duration, split, pad in splits, combine, and implement a Sampler that picks same length samples. It's maybe not ideal as some batches will be smaller and others larger.

The easiest to implement option is to perform padding in collate_fn, but it limits our flexibility in choosing what to pad with, as we're no longer dealing with Cuts at that stage. It also won't help us with dynamic batch size.

I'll prepare an example of an iterable K2 dataset in the next few days.

**Comment by danpovey** [comment]
Mm.  Some datasets will have drastically different lengths of utterance.
If we don't at least try to do something smart on this
(e.g. combine multiple short utterance into a single utterance, with
silence in between so the model can know they're separate),
it will be very inefficient.

**Comment by pzelasko** [comment]
I think after the latest refactoring of all datasets as map-style datasets with cut-aware samplers (#202), this issue can be closed.
